File: ports.cpp
#include "ports.h"
#include <iostream>

Ports::DummyRead Ports::dummyDeviceReadFunction;
Ports::DummyWrite Ports::dummyDeviceWriteFunction;

Ports *Ports::self = NULL;

Ports::Ports()
{
  
}

Ports::~Ports()
{
    std::cout << "Removing all ports" << std::endl;

    delete portDevicesRead;
    delete portDevicesWrite;
    delete portObjectsRead;
    delete portObjectsWrite;
}

Ports *Ports::instance()
{
    if(!self)
    {
        self = new Ports();
        self->init();
    }

    return self;
}

// Allocate the read and write port arrays.
bool Ports::init()
{
    portDevicesRead = new ReadInterface *[MAXPORTS];
    portDevicesWrite = new WriteInterface *[MAXPORTS];

    portObjectsRead = new void *[MAXPORTS];
    portObjectsWrite = new void *[MAXPORTS];
    for (unsigned int i = 0; i < MAXPORTS; i++)
    {
        // Initialise all of the ports to 'dummy', to catch crazy port access.
        portDevicesRead[i] = &dummyDeviceReadFunction;
        portDevicesWrite[i] = &dummyDeviceWriteFunction;
    }

    return true;
}

void Ports::addDeviceToPort(void *object, 
                             ReadInterface *read, 
                             Byte port)

{
    if (read != NULL)
    {
        portDevicesRead[port]  = read;
        portObjectsRead[port] = object;
    }

}

void Ports::addDeviceToPort(void *object, 
                              ReadInterface *read, 
                             WriteInterface *write, 
                             Byte port)
{
    if (read != NULL)
    {
        portDevicesRead[port]  = read;
        portObjectsRead[port] = object;
    }

    if (write != NULL)
        portDevicesWrite[port] = write;

    portObjectsWrite[port] = object;
}

Byte Ports::portRead(Byte port)
{
    return portDevicesRead[port]->read(portObjectsRead[port]);
}

void Ports::portWrite(Byte port, Byte data)
{
    portDevicesWrite[port]->write(portObjectsWrite[port], data);
}

void Ports::portWrite(Byte port, const Byte *data, uint16 length)
{
    portDevicesWrite[port]->write(portObjectsWrite[port], data, length);
}

Byte Ports::dummyDeviceRead(void *object)
{
    std::cout << "Device not implemented, no read" << std::endl;
    return 0;
}

void Ports::dummyDeviceWrite(void *object, Byte data)
{
/*
   std::cout << "Device not implemented (" << 
            (int) data << ") not written" << std::endl;
            */
}

void Ports::dummyDeviceWriteMultiple(void *object, 
                                       const Byte *data, uint16 length)
{
    std::cout << "Multiple write not implemented for this device (" << 
            (int) *data << ") not written" << std::endl;
}

File: pysega/ports.py
import errors

class Port(object):
    def __init__(self, read_func, write_func = None):
        self._write_func = write_func
        self._read_func = read_func

    def read(self):
        return self._read_func()

    def write(self, data):
        self._write_func(data)

class Ports(object):

    MAXPORTS    = 256;

    def __init__(self):
        self.read_ports  = [None] * self.MAXPORTS
        self.write_ports = [None] * self.MAXPORTS

    def addDeviceToPort(self, port_address, read_func, write_func = None):
        if read_func:
          self.read_ports[port_address] = Port(read_func, write_func)

        if write_func:
          self.write_ports[port_address] = Port(read_func, write_func)

    def portRead(self, port_address):
        if (self.read_ports[port_address]):
            return self.read_ports[port_address].read()
        else:
            errors.unsupported("Unsupported port address %s"%(port_address))
            return 0

    def portWrite(self, port_address, value):
        if (self.write_ports[port_address]):
            self.write_ports[port_address].write(value)
        else:
            errors.unsupported("Unsupported port address %s"%(port_address))

    def portMultiWrite(self, port_address, data, length):
        for i in range(length):
            self.portWrite(port_address, data[i])
